06_puppy/lexual/puppy.go

```diff
@@ -0,0 +1,100 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+var out io.Writer = os.Stdout
+
+type Puppy struct {
+	ID     uint32
+	Breed  string
+	Colour string
+	Value  int
+}
+
+// Storer defines standard CRUD operations for Puppys.
+type Storer interface {
+	CreatePuppy(*Puppy)
+	ReadPuppy(id uint32) *Puppy
+	UpdatePuppy(id uint32, puppy *Puppy)
+	DeletePuppy(id uint32) bool
+}
+
+// MapStore ////
+
+type MapStore map[uint32]Puppy
+
+// CreatePuppy stores a copy of Puppy p.
+func (store *MapStore) CreatePuppy(p *Puppy) {
+	(*store)[p.ID] = *p
+}
+
+// ReadPuppy retrieves a previously stored Puppy.
+func (store *MapStore) ReadPuppy(id uint32) *Puppy {
+	p, ok := (*store)[id]
+	if !ok {
+		return nil
+	}
+	return &p
+}
+
+// UpdatePuppy updates details of Puppy in store.
+func (store *MapStore) UpdatePuppy(id uint32, puppy *Puppy) {
+	if id == puppy.ID {
+		(*store)[id] = *puppy
+	}
+}
+
+// DeletePuppy deletes puppy from store.
+func (store *MapStore) DeletePuppy(id uint32) bool {
+	_, exists := (*store)[id]
+	delete(*store, id)
+	return exists
+}
+
+// SyncStore ////
+
+type SyncStore struct {
```

Comment:
I know it wasn't directly stated in the lab description, but another handy facility the `sync` package provides is the Mutex struct.

Up to you whether you choose to do so, it might be fun, but you can integrate Mutex capabilities into your storer implementation.

Example:

```go
// SyncStore implements the Storer interface with Mutex support
type SyncStore struct {
  sync.Map
  sync.Mutex
}

// CreatePuppy creates a new Puppy and loads it into the store
func (s *SyncStore) CreatePuppy(p *Puppy) {
  s.Lock()
  defer s.Unlock()
  s.Store(p.ID, *p)  
}
```